Community-Supported/clouddb-extractor/bigquery_extractor.py

```diff
@@ -0,0 +1,514 @@
+
+    def _hyper_table_definition(self, source_table, hyper_table_name="Extract"):
+        """
+        Build a hyper table definition from source_schema
+
+        source_table (obj): Source table (Instance of google.cloud.bigquery.table.Table)
+        hyper_table_name (string): Name of the target Hyper table, default="Extract"
+
+        Returns a tableauhyperapi.TableDefinition Object
+        """
+
+        logger.debug(
+            "Building Hyper TableDefinition for table {}".format(source_table.reference)
+        )
+        target_cols = []
+        for source_field in source_table.schema:
+            this_name = source_field.name
+            this_type = self._hyper_sql_type(source_field)
+            this_col = TableDefinition.Column(name=this_name, type=this_type)
+
+            # Check for Nullability
+            this_mode = source_field.mode
+            if this_mode == "REPEATED":
+                raise (
+                    HyperSQLTypeMappingError(
+                        "Field mode REPEATED is not implemented in Hyper"
+                    )
+                )
+            if this_mode == "REQUIRED":
+                this_col = TableDefinition.Column(
+                    this_name, this_type, Nullability.NOT_NULLABLE
+                )
+
+            target_cols.append(this_col)
+            logger.debug("..Column {} - Type {}".format(this_name, this_type))
+
+        target_schema = TableDefinition(
+            table_name=TableName("Extract", hyper_table_name), columns=target_cols
+        )
+
+        return target_schema
+
+    def _query_to_hyper_files(self, sql_query, hyper_table_name="Extract"):
+        """
+        Executes sql_query against the source database and writes the output to one or more Hyper files
+        Returns a list of output Hyper files
+
+        sql_query -- SQL string to pass to the source database
+        hyper_table_name -- Name of the target Hyper table, default=Extract
+        """
+
+        # Dry run to estimate result size
+        job_config = bigquery.QueryJobConfig(dry_run=True, use_query_cache=False)
+        dryrun_query_job = bq_client.query(sql_query, job_config=job_config)
+        dryrun_bytes_estimate = dryrun_query_job.total_bytes_processed
+        logger.info("This query will process {} bytes.".format(dryrun_bytes_estimate))
+
+        if dryrun_bytes_estimate > MAX_QUERY_SIZE:
+            raise QuerySizeLimitError(
+                "This query will return than {MAX_QUERY_SIZE} bytes"
+            )
+
+        query_job = bq_client.query(sql_query)
+
+        # Determine table structure
+        query_temp_table = bq_client.get_table(query_job.destination)
+        target_table_def = self._hyper_table_definition(
+            query_temp_table, hyper_table_name
+        )
+
+        def query_job_iter():
+            return bq_client.list_rows(query_job.destination)
+
+        return self._query_result_to_hyper_files(query_job_iter, target_table_def)
```

Comment:
why do we use `_query_result_to_hyper_files` here? Can't we use the same "CSV export" mechanism also for arbitrary SQL queries?

Using the same mechanism to export both SQL queries and raw tables would
1) probably be more efficient for SQL queries
2) allows us to simplify the code significantly, by removing the `_query_result_to_hyper_files` and its helper functions

Reply:
Discussed on slack - can't determine table structure for adhoc queries in the same way as you can for an export of a physical table.  Will get rid of this second mechanism once parquet support is GA in both hyper and BigQuery as then schema will be saved with file - currently uses CSV